libgadget/exchange.c

```diff
@@ -120,13 +92,46 @@ int domain_exchange(int (*layoutfunc)(int p), int do_gc) {
     plan.toGet = (ExchangePlanEntry *) mymalloc2("toGet", sizeof(plan.toGo[0]) * NTask);
     plan.toGetOffset = (ExchangePlanEntry *) mymalloc2("toGet", sizeof(plan.toGo[0]) * NTask);
 
-
     walltime_measure("/Domain/exchange/init");
 
     int iter = 0;
 
     while(MPIU_Any(plan.last < plan.nexchange, MPI_COMM_WORLD))
     {
+        plan.ExchangeList = mymalloc2("exchangelist", sizeof(int) * plan.nexchange * omp_get_max_threads());
```

Comment:
Let's move this chunk of code into a function, e.g. merging the logic into domain_find_iter_space;

Further, ExchangeList, nexchange and last are different from other elements of the plan. We can lift ExchangeList and nexchange (whether a struct or not I don't know yet), which becomes the input argument of domain_build_plan and domain_find_iter_space. `domain_find_iter_space` could have truncated the ExchangeList and flagged that the loop shall not terminate.

Reply:
I'll move it into domain_find_iter_space, but we don't get anything from lifting ExchangeList, nexchange and last into a separate struct because ExchangeList is also used in domain_exchange_once to do the actual copying. So although the constructors are split into dfferent functions, the lifetime of the data is the same.